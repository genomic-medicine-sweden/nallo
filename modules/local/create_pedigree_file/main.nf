process CREATE_PEDIGREE_FILE {
    tag "${meta.id}"
    label 'process_single'

    conda "conda-forge::python=3.8.3"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/python:3.8.3' :
        'biocontainers/python:3.8.3' }"

    input:
    tuple val(meta), val(sample_metas)

    output:
    tuple val(meta), path("*.ped"), emit: ped
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix   = task.ext.prefix ?: "${meta.id}"
    def samples = (sample_metas.collect().size() > 1) ? sample_metas.sort{ a, b ->
        // First sort on family_id, then on sample id
        a.family_id <=> b.family_id ?: a.id <=> b.id } : sample_metas
    outfile_text = ['#family_id', 'sample_id', 'father', 'mother', 'sex', 'phenotype'].join('\\t')
    def samples_list = []
    for(int i = 0; i<samples.size(); i++) {
        sample_name = samples[i].id
        if (!samples_list.contains(sample_name)) {
            outfile_text += "\\n" + [samples[i].family_id, sample_name, samples[i].paternal_id, samples[i].maternal_id, samples[i].sex, samples[i].phenotype].join('\\t')
            samples_list.add(sample_name)
        }
    }
    """
    echo -e "$outfile_text" > ${prefix}.ped

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        create_pedigree_file: 1.0
        python: \$(python --version | sed 's/Python //g')
    END_VERSIONS
    """

    stub:
    def prefix   = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.ped

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        create_pedigree_file: 1.0
        python: \$(python --version | sed 's/Python //g')
    END_VERSIONS
    """
}
