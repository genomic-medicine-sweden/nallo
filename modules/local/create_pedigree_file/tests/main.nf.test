nextflow_process {

    name "Test Process CREATE_PEDIGREE_FILE"
    script "../main.nf"
    process "CREATE_PEDIGREE_FILE"

    tag "modules"
    tag "modules_local"
    tag "create_pedigree_file"

    test("create_pedigree_file - one sample") {

        when {
            process {
                """
                input[0] = Channel.of(
                    [ 'project',
                    [
                        [id:'HG002.Revio', family_id:'family', paternal_id:0, maternal_id:0, sex:0, phenotype:2, project:'project', n_files:1, single_end:true]
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    path(process.out.ped.get(0).get(1)).readLines(),
                    process.out.versions
                ).match() }
            )
        }
    }

    test("create_pedigree_file - multiple samples") {

        when {
            process {
                """
                input[0] = Channel.of(
                    [ 'test',
                    [
                        [id:'HG002_Revio_AB', family_id:'FAM2', paternal_id:0, maternal_id:0, sex:0, phenotype:1, project:'test', n_files:1, single_end:true],
                        [id:'HG002_Revio_AA', family_id:'FAM2', paternal_id:0, maternal_id:0, sex:0, phenotype:1, project:'test', n_files:1, single_end:true],
                        [id:'HG002_Revio_B', family_id:'FAM1', paternal_id:0, maternal_id:0, sex:0, phenotype:1, project:'test', n_files:2, single_end:true],
                        [id:'HG002_Revio_A', family_id:'FAM1', paternal_id:0, maternal_id:0, sex:0, phenotype:2, project:'test', n_files:1, single_end:true],
                        [id:'HG002_Revio_B', family_id:'FAM1', paternal_id:0, maternal_id:0, sex:0, phenotype:1, project:'test', n_files:2, single_end:true]
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    path(process.out.ped.get(0).get(1)).readLines(),
                    process.out.versions
                ).match() }
            )
        }
    }

}
