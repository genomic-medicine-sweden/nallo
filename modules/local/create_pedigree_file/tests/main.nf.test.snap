{
    "create_pedigree_file - one sample": {
        "content": [
            [
                "#family_id\tsample_id\tfather\tmother\tsex\tphenotype",
                "family\tHG002.Revio\t0\t0\t0\t2"
            ],
            [
                "versions.yml:md5,c0d2b2d33f1c465351d65e149ae88f0c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-09-12T10:24:34.60478278"
    },
    "create_pedigree_file - multiple samples": {
        "content": [
            [
                "#family_id\tsample_id\tfather\tmother\tsex\tphenotype",
                "FAM1\tHG002_Revio_A\t0\t0\t0\t2",
                "FAM1\tHG002_Revio_B\t0\t0\t0\t1",
                "FAM2\tHG002_Revio_AA\t0\t0\t0\t1",
                "FAM2\tHG002_Revio_AB\t0\t0\t0\t1"
            ],
            [
                "versions.yml:md5,c0d2b2d33f1c465351d65e149ae88f0c"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-09-12T10:40:54.71628575"
    }
}